File: backend/app/workouts/exercise_counter.py
"""
Exercise counter with angle-based detection for workout tracking.
Adapted from Good-GYM-master for FastAPI backend.
"""
import numpy as np
from collections import deque
import time
import json
import os
from typing import Optional, Dict, Any, List, Tuple


class ExerciseCounter:
    """Basic exercise counter with angle-based detection"""

    def __init__(self, exercises_config_path: str, smoothing_window: int = 5):
        # Core counting variables
        self.counter = 0
        self.stage = None

        # Basic features
        self.smoothing_window = smoothing_window
        self.angle_history = deque(maxlen=smoothing_window)
        self.last_count_time = 0
        self.min_rep_time = 0.5  # Minimum time between reps (seconds)

        # Form corrections
        self.form_corrections = []
        self.last_angle = None

        # Exercise configurations
        self.exercise_configs = self.load_exercise_configs(exercises_config_path)

        # Independent counting for leg exercises - load from config
        self.leg_exercises = [
            exercise_type for exercise_type, config in self.exercise_configs.items()
            if config.get('is_leg_exercise', False)
        ]
        self.leg_stages = {'left': None, 'right': None}  # Track each leg's stage

    def load_exercise_configs(self, config_path: str) -> Dict[str, Any]:
        """Load exercise-specific angle thresholds from JSON file"""
        try:
            if os.path.exists(config_path):
                with open(config_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    exercises = data.get('exercises', {})

                    # Convert to the format expected by the rest of the code
                    configs = {}
                    for exercise_type, config in exercises.items():
                        configs[exercise_type] = {
                            'down_angle': config.get('down_angle'),
                            'up_angle': config.get('up_angle'),
                            'keypoints': config.get('keypoints', {}),
                            'is_leg_exercise': config.get('is_leg_exercise', False)
                        }

                    print(f"✓ Loaded {len(configs)} exercises from {config_path}")
                    return configs
            else:
                print(f"ERROR: Exercises file not found at {config_path}")
                return {}
        except Exception as e:
            print(f"ERROR loading exercises from JSON: {e}")
            return {}

    def reset_counter(self):
        """Reset counter to initial state"""
        self.counter = 0
        self.stage = None
        self.angle_history.clear()
        self.leg_stages = {'left': None, 'right': None}
        self.form_corrections = []
        self.last_angle = None

    def calculate_angle(self, a: np.ndarray, b: np.ndarray, c: np.ndarray) -> Optional[float]:
        """Calculate angle between three points"""
        try:
            a = np.array(a, dtype=np.float64)
            b = np.array(b, dtype=np.float64)
            c = np.array(c, dtype=np.float64)

            # Check for invalid points
            if np.any(np.isnan([a, b, c])) or np.any([a, b, c] == [0, 0]):
                return None

            # Calculate vectors
            ba = a - b
            bc = c - b

            # Check for zero vectors
            ba_norm = np.linalg.norm(ba)
            bc_norm = np.linalg.norm(bc)

            if ba_norm == 0 or bc_norm == 0:
                return None

            # Calculate angle using dot product
            cosine_angle = np.dot(ba, bc) / (ba_norm * bc_norm)
            cosine_angle = np.clip(cosine_angle, -1.0, 1.0)  # Prevent numerical errors
            angle = np.arccos(cosine_angle)

            return float(np.degrees(angle))

        except Exception as e:
            print(f"Angle calculation error: {e}")
            return None

    def smooth_angle(self, angle: Optional[float]) -> Optional[float]:
        """Apply smoothing to reduce noise"""
        if angle is None:
            return None

        self.angle_history.append(angle)

        if len(self.angle_history) < 3:
            return angle

        # Use median filter to remove outliers, then average
        angles_array = np.array(list(self.angle_history))
        median_angle = np.median(angles_array)

        # Remove outliers (angles > 2 std devs from median)
        std_dev = np.std(angles_array)
        filtered_angles = angles_array[np.abs(angles_array - median_angle) <= 2 * std_dev]

        return float(np.mean(filtered_angles)) if len(filtered_angles) > 0 else angle

    def check_rep_timing(self) -> bool:
        """Prevent counting reps too quickly"""
        current_time = time.time()
        if current_time - self.last_count_time < self.min_rep_time:
            return False
        return True

    def count_exercise(self, keypoints: np.ndarray, exercise_type: str) -> Optional[float]:
        """Generic exercise counting function"""
        try:
            if exercise_type not in self.exercise_configs:
                print(f"Unknown exercise type: {exercise_type}")
                return None

            config = self.exercise_configs[exercise_type]
            kp = config['keypoints']

            # Calculate angles for both sides
            left_angle = self.calculate_angle(
                keypoints[kp['left'][0]],  # first point
                keypoints[kp['left'][1]],  # middle point
                keypoints[kp['left'][2]]   # last point
            )

            right_angle = self.calculate_angle(
                keypoints[kp['right'][0]],  # first point
                keypoints[kp['right'][1]],  # middle point
                keypoints[kp['right'][2]]   # last point
            )

            if left_angle is None or right_angle is None:
                return None

            # Handle leg exercises differently
            if exercise_type in self.leg_exercises:
                return self.count_leg_exercise(left_angle, right_angle, config)

            # For other exercises, use average angle
            avg_angle = (left_angle + right_angle) / 2
            smoothed_angle = self.smooth_angle(avg_angle)

            if smoothed_angle is None:
                return None

            # Get thresholds
            up_threshold = config['up_angle']
            down_threshold = config['down_angle']

            # Check form quality
            self._check_form_quality(smoothed_angle, exercise_type, up_threshold, down_threshold)

            # Counting logic with timing check
            if smoothed_angle > up_threshold:
                self.stage = "up"
            elif (smoothed_angle < down_threshold and
                  self.stage == "up" and
                  self.check_rep_timing()):

                self.stage = "down"
                self.counter += 1
                self.last_count_time = time.time()

            return smoothed_angle

        except Exception as e:
            print(f"Exercise counting error: {e}")
            return None

    def count_leg_exercise(self, left_angle: float, right_angle: float, config: Dict[str, Any]) -> float:
        """Count leg exercises with complete up-down cycles"""
        up_threshold = config['up_angle']
        down_threshold = config['down_angle']

        # Check if either leg meets the criteria
        if self.check_rep_timing():
            # Left leg
            if left_angle > up_threshold:
                self.leg_stages['left'] = "up"
            elif (left_angle < down_threshold and
                  self.leg_stages['left'] == "up"):
                self.counter += 1
                self.last_count_time = time.time()
                self.leg_stages['left'] = "down"

            # Right leg
            if right_angle > up_threshold:
                self.leg_stages['right'] = "up"
            elif (right_angle < down_threshold and
                  self.leg_stages['right'] == "up"):
                self.counter += 1
                self.last_count_time = time.time()
                self.leg_stages['right'] = "down"

        # Return average angle for display purposes
        return (left_angle + right_angle) / 2

    # Wrapper functions for different exercises
    def count_squat(self, keypoints: np.ndarray) -> Optional[float]:
        """Count squat repetitions"""
        return self.count_exercise(keypoints, 'squat')

    def count_pushup(self, keypoints: np.ndarray) -> Optional[float]:
        """Count pushup repetitions"""
        return self.count_exercise(keypoints, 'pushup')

    def count_situp(self, keypoints: np.ndarray) -> Optional[float]:
        """Count situp repetitions"""
        return self.count_exercise(keypoints, 'situp')

    def count_bicep_curl(self, keypoints: np.ndarray) -> Optional[float]:
        """Count bicep curl repetitions"""
        return self.count_exercise(keypoints, 'bicep_curl')

    def count_lateral_raise(self, keypoints: np.ndarray) -> Optional[float]:
        """Count lateral raise repetitions"""
        return self.count_exercise(keypoints, 'lateral_raise')

    def count_overhead_press(self, keypoints: np.ndarray) -> Optional[float]:
        """Count overhead press repetitions"""
        return self.count_exercise(keypoints, 'overhead_press')

    def count_leg_raise(self, keypoints: np.ndarray) -> Optional[float]:
        """Count leg raise repetitions"""
        return self.count_exercise(keypoints, 'leg_raise')

    def count_knee_raise(self, keypoints: np.ndarray) -> Optional[float]:
        """Count knee raise repetitions"""
        return self.count_exercise(keypoints, 'knee_raise')

    def count_knee_press(self, keypoints: np.ndarray) -> Optional[float]:
        """Count knee press repetitions"""
        return self.count_exercise(keypoints, 'knee_press')

    def count_crunch(self, keypoints: np.ndarray) -> Optional[float]:
        """Count crunch repetitions"""
        return self.count_exercise(keypoints, 'crunch')

    def get_counter(self) -> int:
        """Get current rep count"""
        return self.counter

    def get_stage(self) -> Optional[str]:
        """Get current exercise stage"""
        return self.stage

    def get_form_corrections(self) -> List[str]:
        """Return and reset form corrections"""
        corrections = self.form_corrections.copy()
        self.form_corrections = []
        return corrections

    def _check_form_quality(
        self,
        angle: float,
        exercise_type: str,
        up_threshold: float,
        down_threshold: float
    ):
        """
        Analyze form and add corrections if needed.

        Args:
            angle: Current angle measurement
            exercise_type: Type of exercise being performed
            up_threshold: Upper angle threshold for exercise
            down_threshold: Lower angle threshold for exercise
        """

        # Squat-specific corrections
        if exercise_type == "squat":
            if self.stage == 'down' and angle > down_threshold + 20:
                self.form_corrections.append("Присядьте глубже - колени должны быть под углом 90°")
            elif self.stage == 'up' and angle < up_threshold - 10:
                self.form_corrections.append("Полностью выпрямите ноги")

        # Pushup-specific corrections
        elif exercise_type == "pushup":
            if self.stage == 'down' and angle > down_threshold + 15:
                self.form_corrections.append("Опуститесь ниже - грудь ближе к полу")
            elif self.stage == 'up' and angle < up_threshold - 15:
                self.form_corrections.append("Полностью выпрямите руки")

        # Lunge-specific corrections
        elif exercise_type == "lunge":
            if self.stage == 'down' and angle > down_threshold + 20:
                self.form_corrections.append("Опустите колено ниже - угол 90°")

        # General correction for insufficient range of motion
        if self.last_angle is not None:
            angle_change = abs(angle - self.last_angle)
            if angle_change < 5 and self.stage is not None:
                self.form_corrections.append("Увеличьте амплитуду движения")

        self.last_angle = angle

File: dashboard/app/utils/exerciseCounter.ts
/**
 * Exercise counter with angle-based detection for workout tracking.
 * TypeScript port of the Python exercise_counter.py
 */

import type {
  ExerciseConfig,
  ExerciseStage,
  LegStages,
  PoseKeypoints
} from '~/types/pose';
import exercisesData from '~/data/exercises.json';

// Load exercise configurations - cast to proper type
const exercisesConfig = exercisesData as { exercises: Record<string, ExerciseConfig> };

export class ExerciseCounter {
  // Core counting variables
  private counter: number = 0;
  private stage: ExerciseStage = null;

  // Smoothing
  private smoothingWindow: number;
  private angleHistory: number[] = [];
  private lastCountTime: number = 0;
  private minRepTime: number = 500; // Minimum time between reps (milliseconds)

  // Form corrections
  private formCorrections: string[] = [];
  private lastAngle: number | null = null;

  // Exercise configurations
  private exerciseConfigs: Record<string, ExerciseConfig>;

  // Independent counting for leg exercises
  private legExercises: string[];
  private legStages: LegStages = { left: null, right: null };

  // Timed exercises (like plank)
  private timedExercises: string[];
  private isInPosition: boolean = false;

  constructor(smoothingWindow: number = 5) {
    this.smoothingWindow = smoothingWindow;
    this.exerciseConfigs = exercisesConfig.exercises;

    // Load leg exercises from config
    this.legExercises = Object.entries(this.exerciseConfigs)
      .filter(([_, config]) => config.is_leg_exercise)
      .map(([type]) => type);

    // Load timed exercises from config
    this.timedExercises = Object.entries(this.exerciseConfigs)
      .filter(([_, config]) => config.is_timed_exercise)
      .map(([type]) => type);
  }

  /**
   * Reset counter to initial state
   */
  resetCounter(): void {
    this.counter = 0;
    this.stage = null;
    this.angleHistory = [];
    this.legStages = { left: null, right: null };
    this.formCorrections = [];
    this.lastAngle = null;
    this.lastCountTime = 0;
    this.isInPosition = false;
  }

  /**
   * Calculate angle between three points
   */
  calculateAngle(
    a: number[] | undefined,
    b: number[] | undefined,
    c: number[] | undefined
  ): number | null {
    try {
      // Check for invalid points
      if (!a || !b || !c || a.length < 2 || b.length < 2 || c.length < 2) {
        return null;
      }

      const a0 = a[0] ?? 0;
      const a1 = a[1] ?? 0;
      const b0 = b[0] ?? 0;
      const b1 = b[1] ?? 0;
      const c0 = c[0] ?? 0;
      const c1 = c[1] ?? 0;

      if (
        isNaN(a0) || isNaN(a1) ||
        isNaN(b0) || isNaN(b1) ||
        isNaN(c0) || isNaN(c1) ||
        (a0 === 0 && a1 === 0) ||
        (b0 === 0 && b1 === 0) ||
        (c0 === 0 && c1 === 0)
      ) {
        return null;
      }

      // Calculate vectors
      const baX = a0 - b0;
      const baY = a1 - b1;
      const bcX = c0 - b0;
      const bcY = c1 - b1;

      // Calculate magnitudes
      const baNorm = Math.sqrt(baX * baX + baY * baY);
      const bcNorm = Math.sqrt(bcX * bcX + bcY * bcY);

      if (baNorm === 0 || bcNorm === 0) {
        return null;
      }

      // Calculate angle using dot product
      const dotProduct = baX * bcX + baY * bcY;
      let cosineAngle = dotProduct / (baNorm * bcNorm);

      // Clamp to prevent numerical errors
      cosineAngle = Math.max(-1, Math.min(1, cosineAngle));
      const angle = Math.acos(cosineAngle);

      return angle * (180 / Math.PI);
    } catch (error) {
      console.error('Angle calculation error:', error);
      return null;
    }
  }

  /**
   * Apply smoothing to reduce noise using median filter
   */
  smoothAngle(angle: number | null): number | null {
    if (angle === null) {
      return null;
    }

    this.angleHistory.push(angle);

    // Keep only the last N angles
    if (this.angleHistory.length > this.smoothingWindow) {
      this.angleHistory.shift();
    }

    if (this.angleHistory.length < 3) {
      return angle;
    }

    // Use median filter to remove outliers, then average
    const sorted = [...this.angleHistory].sort((a, b) => a - b);
    const medianAngle = sorted[Math.floor(sorted.length / 2)] ?? angle;

    // Calculate standard deviation
    const mean = this.angleHistory.reduce((a, b) => a + b, 0) / this.angleHistory.length;
    const stdDev = Math.sqrt(
      this.angleHistory.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) /
      this.angleHistory.length
    );

    // Filter outliers (angles > 2 std devs from median)
    const filteredAngles = this.angleHistory.filter(
      a => Math.abs(a - medianAngle) <= 2 * stdDev
    );

    if (filteredAngles.length === 0) {
      return angle;
    }

    return filteredAngles.reduce((a, b) => a + b, 0) / filteredAngles.length;
  }

  /**
   * Prevent counting reps too quickly
   */
  private checkRepTiming(): boolean {
    const currentTime = Date.now();
    if (currentTime - this.lastCountTime < this.minRepTime) {
      return false;
    }
    return true;
  }

  /**
   * Generic exercise counting function
   */
  countExercise(
    keypoints: PoseKeypoints,
    exerciseType: string
  ): number | null {
    try {
      const config = this.exerciseConfigs[exerciseType];
      if (!config) {
        console.warn(`Unknown exercise type: ${exerciseType}`);
        return null;
      }

      const kp = config.keypoints;

      // Validate keypoint indices exist
      if (kp.left.length < 3 || kp.right.length < 3) {
        return null;
      }

      const leftIdx0 = kp.left[0]!;
      const leftIdx1 = kp.left[1]!;
      const leftIdx2 = kp.left[2]!;
      const rightIdx0 = kp.right[0]!;
      const rightIdx1 = kp.right[1]!;
      const rightIdx2 = kp.right[2]!;

      // Calculate angles for both sides
      const leftAngle = this.calculateAngle(
        keypoints[leftIdx0] ?? undefined,
        keypoints[leftIdx1] ?? undefined,
        keypoints[leftIdx2] ?? undefined
      );

      const rightAngle = this.calculateAngle(
        keypoints[rightIdx0] ?? undefined,
        keypoints[rightIdx1] ?? undefined,
        keypoints[rightIdx2] ?? undefined
      );

      if (leftAngle === null || rightAngle === null) {
        return null;
      }

      // Handle leg exercises differently
      if (this.legExercises.includes(exerciseType)) {
        return this.countLegExercise(leftAngle, rightAngle, config);
      }

      // For other exercises, use average angle
      const avgAngle = (leftAngle + rightAngle) / 2;
      const smoothedAngle = this.smoothAngle(avgAngle);

      if (smoothedAngle === null) {
        return null;
      }

      // Get thresholds
      const upThreshold = config.up_angle;
      const downThreshold = config.down_angle;

      // Check form quality
      this.checkFormQuality(smoothedAngle, exerciseType, upThreshold, downThreshold);

      // Counting logic with timing check
      if (smoothedAngle > upThreshold) {
        this.stage = 'up';
      } else if (
        smoothedAngle < downThreshold &&
        this.stage === 'up' &&
        this.checkRepTiming()
      ) {
        this.stage = 'down';
        this.counter++;
        this.lastCountTime = Date.now();
      }

      return smoothedAngle;
    } catch (error) {
      console.error('Exercise counting error:', error);
      return null;
    }
  }

  /**
   * Count leg exercises with complete up-down cycles for each leg independently
   */
  private countLegExercise(
    leftAngle: number,
    rightAngle: number,
    config: ExerciseConfig
  ): number {
    const upThreshold = config.up_angle;
    const downThreshold = config.down_angle;

    if (this.checkRepTiming()) {
      // Left leg
      if (leftAngle > upThreshold) {
        this.legStages.left = 'up';
      } else if (
        leftAngle < downThreshold &&
        this.legStages.left === 'up'
      ) {
        this.counter++;
        this.lastCountTime = Date.now();
        this.legStages.left = 'down';
      }

      // Right leg
      if (rightAngle > upThreshold) {
        this.legStages.right = 'up';
      } else if (
        rightAngle < downThreshold &&
        this.legStages.right === 'up'
      ) {
        this.counter++;
        this.lastCountTime = Date.now();
        this.legStages.right = 'down';
      }
    }

    // Return average angle for display purposes
    return (leftAngle + rightAngle) / 2;
  }

  /**
   * Check form quality and add corrections if needed
   */
  private checkFormQuality(
    angle: number,
    exerciseType: string,
    upThreshold: number,
    downThreshold: number
  ): void {
    // Squat-specific corrections
    if (exerciseType === 'squat') {
      if (this.stage === 'down' && angle > downThreshold + 20) {
        this.formCorrections.push('Присядьте глубже - колени должны быть под углом 90°');
      } else if (this.stage === 'up' && angle < upThreshold - 10) {
        this.formCorrections.push('Полностью выпрямите ноги');
      }
    }
    // Pushup-specific corrections
    else if (exerciseType === 'pushup') {
      if (this.stage === 'down' && angle > downThreshold + 15) {
        this.formCorrections.push('Опуститесь ниже - грудь ближе к полу');
      } else if (this.stage === 'up' && angle < upThreshold - 15) {
        this.formCorrections.push('Полностью выпрямите руки');
      }
    }
    // Lunge-specific corrections
    else if (exerciseType === 'lunge') {
      if (this.stage === 'down' && angle > downThreshold + 20) {
        this.formCorrections.push('Опустите колено ниже - угол 90°');
      }
    }

    // General correction for insufficient range of motion
    if (this.lastAngle !== null) {
      const angleChange = Math.abs(angle - this.lastAngle);
      if (angleChange < 5 && this.stage !== null) {
        this.formCorrections.push('Увеличьте амплитуду движения');
      }
    }

    this.lastAngle = angle;
  }

  /**
   * Get current rep count
   */
  getCounter(): number {
    return this.counter;
  }

  /**
   * Get current exercise stage
   */
  getStage(): ExerciseStage {
    return this.stage;
  }

  /**
   * Return and reset form corrections
   */
  getFormCorrections(): string[] {
    const corrections = [...this.formCorrections];
    this.formCorrections = [];
    return corrections;
  }

  /**
   * Get exercise configuration
   */
  getExerciseConfig(exerciseType: string): ExerciseConfig | undefined {
    return this.exerciseConfigs[exerciseType];
  }

  /**
   * Check if exercise type exists
   */
  hasExercise(exerciseType: string): boolean {
    return exerciseType in this.exerciseConfigs;
  }

  /**
   * Check if exercise is a timed exercise (like plank)
   */
  isTimedExercise(exerciseType: string): boolean {
    return this.timedExercises.includes(exerciseType);
  }

  /**
   * Check if user is in correct position for timed exercise
   */
  getIsInPosition(): boolean {
    return this.isInPosition;
  }

  /**
   * Check timed exercise position (e.g., plank)
   * Returns the current angle if in position, null otherwise
   */
  checkTimedExercise(
    keypoints: PoseKeypoints,
    exerciseType: string
  ): { inPosition: boolean; angle: number | null } {
    const config = this.exerciseConfigs[exerciseType];
    if (!config) {
      return { inPosition: false, angle: null };
    }

    const kp = config.keypoints;

    // Validate keypoint indices exist
    if (kp.left.length < 3 || kp.right.length < 3) {
      return { inPosition: false, angle: null };
    }

    const leftIdx0 = kp.left[0]!;
    const leftIdx1 = kp.left[1]!;
    const leftIdx2 = kp.left[2]!;
    const rightIdx0 = kp.right[0]!;
    const rightIdx1 = kp.right[1]!;
    const rightIdx2 = kp.right[2]!;

    // Calculate angles for both sides
    const leftAngle = this.calculateAngle(
      keypoints[leftIdx0] ?? undefined,
      keypoints[leftIdx1] ?? undefined,
      keypoints[leftIdx2] ?? undefined
    );

    const rightAngle = this.calculateAngle(
      keypoints[rightIdx0] ?? undefined,
      keypoints[rightIdx1] ?? undefined,
      keypoints[rightIdx2] ?? undefined
    );

    if (leftAngle === null || rightAngle === null) {
      this.isInPosition = false;
      return { inPosition: false, angle: null };
    }

    const avgAngle = (leftAngle + rightAngle) / 2;
    const smoothedAngle = this.smoothAngle(avgAngle);

    if (smoothedAngle === null) {
      this.isInPosition = false;
      return { inPosition: false, angle: null };
    }

    // For plank: user should be in a straight position
    // Angle should be between down_angle and up_angle
    const inPosition = smoothedAngle >= config.down_angle && smoothedAngle <= config.up_angle;
    this.isInPosition = inPosition;

    // Add form correction if not in position
    if (!inPosition) {
      if (smoothedAngle < config.down_angle) {
        this.formCorrections.push('Выпрямите тело - не провисайте');
      } else if (smoothedAngle > config.up_angle) {
        this.formCorrections.push('Опустите бёдра - тело должно быть прямым');
      }
    }

    return { inPosition, angle: Math.round(smoothedAngle) };
  }
}
